faq.md

```diff
@@ -575,6 +575,144 @@ If you are using the Flutter plugin for IntelliJ, you can launch the app in
+first describing the situation under test, then listing the invariants that the
+test must match, then advancing the clock or inserting events as necessary for
+the test.
+
+* Reactive programming: The widget and element trees are sometimes described as
+reactive, because new inputs provided in a widget's constructor are immediately
+propagated as changes to lower-level widgets by the widget's build method, and
+changes made in the lower widgets (e.g. in response to user input) propagate
+back up the tree via event handlers. Aspects of both functional-reactive and
+imperative-reactive are present in the framework, depending on the needs of the
+widgets. For example, the material
+[Divider](https://docs.flutter.io/flutter/material/Divider-class.html) class has
+a build method that consists of just an expression describing how this widget
+reacts to changes in its configuration (a functional reactive widget), while the
+[Chip](https://docs.flutter.io/flutter/material/Chip-class.html) class has a
+build method that constructs a list of children over several statements,
+describing how this widget reacts to changes in its configuration (an imperative
+reactive widget).
+
+* Declarative programming: The build methods of widgets are often a single
+expression with multiple levels of nested constructors, written using a strictly
+declarative subset of Dart (such nested expressions could be mechanically
+transformed to or from any suitably expressive markup language). For example,
+the
+[UserAccountsDrawerHeader](https://docs.flutter.io/flutter/material/UserAccountsDrawerHeader-class.html)
+widget has a long build method (20+ lines), consisting of a single nested
+expression. This can also be combined with the imperative style to build UIs
+that would be harder to describe in a pure-declarative approach.
+
+* Generic programming: Types can be used to help catch programming errors early.
+The Flutter framework uses generic programming to help in this regard. For
+example, the [State](https://docs.flutter.io/flutter/widgets/State-class.html)
+class is parameterized in terms of the type of its associated widget, so that
+the Dart analyzer can catch mismatches of states and widgets. Similarly, the
+[GlobalKey](https://docs.flutter.io/flutter/widgets/GlobalKey-class.html) class
+takes a type parameter so that it can access a remote widget's state in a
+type-safe manner (with runtime checking), the
+[Route](https://docs.flutter.io/flutter/widgets/Route-class.html) interface is
+parameterized with the type that it is expected to use when
+[popped](https://docs.flutter.io/flutter/widgets/Navigator/pop.html), and
+collections such as
+[List](https://docs.flutter.io/flutter/dart-core/List-class.html)s,
+[Map](https://docs.flutter.io/flutter/dart-core/Map-class.html)s, and
+[Set](https://docs.flutter.io/flutter/dart-core/Set-class.html)s are all
+parameterized so that mismatched elements can be caught early either by the
```

Comment:
"can be caught during debugging by either the analyzer or at runtime."

Reply:
changed to "either during analysis or at runtime during debugging". The analyzer isn't part of debugging.